library/minecraft_server_file.py

```diff
@@ -0,0 +1,354 @@
+    REQUIRED_ARGS = ['owner', 'group']
+
+    def __init__(self, module):
+        self.module = module
+        missing_args = [x for x in self.REQUIRED_ARGS
+                        if not self.module.params[x]]
+        if missing_args:
+            raise MissingArgsException(missing_args)
+
+        self.path = os.path.expanduser(module.params['path'])
+        self.owner = module.params['owner']
+        self.group = module.params['group']
+        self.mode = module.params['mode'] if module.params['mode'] else 0644
+
+    def file_stats_changed(self):
+        try:
+            st = os.stat(self.path)
+        except OSError:
+            # If the file doesn't exist or isn't read, this is the same as
+            # the stat values being wrong.
+            return True
+
+        current_mode = st.st_mode & stat.S_IMODE(st.st_mode)
+        # We cannot look up non-existant uids and gids when we compare the
+        # current state to desired state. Instead, get the usernames and
+        # groups now to make comparison easier.
+        current_uid, current_gid = self.module.user_and_group(self.path)
+        current_user = pwd.getpwuid(current_uid).pw_name
+        current_group = grp.getgrgid(current_gid).gr_name
+
+        return any([
+            self.owner != current_user,
+            self.group != current_group,
+            self.mode != current_mode,
+        ])
+
+    def correct_file_stats(self):
+        self.module.set_owner_if_different(self.path, self.owner, False)
+        self.module.set_group_if_different(self.path, self.group, False)
+        self.module.set_mode_if_different(self.path, self.mode, False)
+
+
+class ServerFile(object):
+    """
+    Abstract class representing any server file.
+    """
+    def __init__(self, module):
+        self.module = module
+        self.values = self.module.params['values']
+        self.stats = FileStats(module)
+
+    def content(self):
+        """
+        Return content in its native format (e.g. dict).
+        """
+        raise NotImplementedError
+
+    def content_str(self):
+        """
+        Return content as a string, for saving to a file.
+        """
+        return self.content()
+
+    def content_changed(self):
+        """
+        Indicates whether the content changed or not.
+        """
+        raise NotImplementedError
+
+
+class ServerProperties(ServerFile):
+    """
+    The Minecraft server.properties file.
+
+    The module "values" argument should be a list of strings in the format
+    "<name>=<value>".
+    """
+    def __init__(self, module):
+        super(ServerProperties, self).__init__(module)
+        self._content_changed = False
+        self.newlines = []
+        properties = {}
+
+        for a in self.values:
+            name, eq, val = a.partition('=')
+            if not eq:
+                raise BadValueException(name)
+            properties[name] = val
+
+        filein = file(self.stats.path)
+
+        for line in filein:
```

Comment:
I made the change you suggested, but upon testing and reading more about .properties files, it feels like an even hackier solution now.

As far as I can tell, the values in a Java properties file are actually always treated as strings (at least that's what I gleaned from https://docs.oracle.com/javase/tutorial/essential/environment/properties.html. I've done very little Java coding myself.) But if we use a dict in the yml file, they'll be translated into Python types. This is okay for strings and integers, but when writing them out booleans, for example, they will appear as "True" (the literal string version of the boolean `True` value) instead of "true". I don't know that this is actually a problem, but it concerns me that there might be other cases of nonstandard translation. Or that we have to enforce that people quote the yml dict values, and that doesn't seem user friendly.

Furthermore, it looks like .properties files support even more options than ConfigParser can handle (such as `key value`, and backslash escapes).

I'm starting to think that this solution is more trouble than it's worth, since we can't really get to a proper solution. ConfigParser seems to buy us only two things: the ability to use a dict-type object, and support for `:` instead of `=` as separator. If we think that the default file will stick with simple `foo=bar` declarations, and if the intention is that customizations to that file on a given server occur only via this ansible role, then having the user provide that style of entry in their playbook seems like the simplest solution and fairly safe—particularly since most users have no idea what a Java properties file is. They just know that they have to add "enable-rcon=true" to that file to enable rcon, for example. I could (and probably should) also hedge slightly and at least trim whitespace off the parsed values.

What do you think?